### ISSUE TITLE
Tab-index, pushing affordance, and hover styling for dialog buttons

### ISSUE THREAD

**Issue by ivanov** [jupyterlab/jupyterlab#6230]
**Describe the bug**
Insufficient styling for dialog buttons.

For example, the "File Changed" dialog (see #6229 for steps to trigger this dialog) does have tab-indexing, but it can only be seen (as of 0.35.4) in Firefox - chrome does not show an outline for the three options as you switch between them.

Example in Chrome (hitting tab does not show an outline on the buttons):
![image](https://user-images.githubusercontent.com/118211/56532929-975cd580-650b-11e9-950f-73f5f7d93c4c.png)

Same example in Firefox:
![image](https://user-images.githubusercontent.com/118211/56532967-a5125b00-650b-11e9-8997-78fc33d87ad6.png)

**Expected behavior**
All dialog buttons should contain a stylized tab-index indicator, a pushing affordance (some notion of button being clicked and triggering an action) and also hover styling for mouse-over indication of it being a clickable button.

- JupyterLab 0.35.4

**Related bugs**
#3473 is about having a tab-index and a sensible default, this bug is about styling the tab-index indicator

**Comment by jasongrout** [comment]
I'll note that I saw the same focus highlight in chrome and firefox mentioned above on current master.

**Comment by ellisonbg** [comment]
Yes, let's do this for 1.0. Getting this right is challenging, but will be worth it.

**Comment by ellisonbg** [comment]
We are using Blueprint buttons now, and it addresses these things:

https://blueprintjs.com/docs/#core/components/button

I don't think we want to use all of their styling (they use gradients), but it is a starting point.

**Comment by saulshanabrook** [comment]
The challenging part is that there are a lot of different button states. The visual representations all have to be different. It's hard to have them all look different without some of them looking bad. So we should be careful when fixing this not to add bad UX.

**Comment by ellisonbg** [comment]
K, working on this....buttons on the web are a mess.

active = state when clicked
focus = state when button has focus
hover = state when mouse is hovering over element

Here is a summary of how Blueprint, MD, and Fabric handle these things:

* Blueprint
  - hover: light gradient
  - active: darker bg
  - focus: offset exterior outline
* Material Design
  - hover: raise + lighten bg
  - active: expanding disc animation
  - focus: raise + lighten bg (subtly different from hover)
* Fabric
  - hover: darker bg
  - active: darker still
  - focus: inset outline

Material Design is the most odd of these, some combination of the other two is probably closest to where we are going. I will iterate and play with the options.

**Comment by jasongrout** [comment]
One of the things originating this conversation was multiple identical dialogs on top of each other, so that pushing a button didn't seem to do anything. Even with button affordances, it will seem that the button can be pushed, but the dialog does nothing (but in reality it goes away and the identical dialog behind it appears). Addressing this, we can:

1. make it so that identical dialogs don't pop up (i.e., somehow the dialog creation call realizes that there already is a dialog up, and errors, or returns the existing dialog promise, or something).
2. Make some indication that the dialog is going away, like fading away, collapsing, or something. That gives the user an indication that the dialog they were just looking at went away, and they are now looking a different (but identical) dialog.

**Comment by ellisonbg** [comment]
I wasn't aware of the multiple identical dialog issue, that will certainly complicate this. To keep making progress, I am going to ignore that and focus on the core button UX.

I have been diving into things and am finding this pretty tedious. Some of the challenges:

* Even though we have a button react component, we are not using it very many places, so there are lots of places to fix this. As a minimum, to fix this issue in the dialog we need to work on the CSS for the dialog buttons, which use the old `jp-mod-styled` class. Unfortunately, none of the other buttons will get the benefit of the improvements.
* Blue solid background buttons are the most challenging to handle focus on from a design perspective because the canonical focus color is also blue.
* All our buttons are wrapped in Widgets that set `overflow: hidden`, so any exterior outline based solution won't work without starting to change lots of other parent containers (the outlines are visually truncated). For the dialog this isn't too bad, but for toolbars that are compact, it involves a number of parents where they would be unwanted side effects to changing to `overflow:visible`.
* From a personal psychological perspective I have a tough time fixing a mess by making more of a mess, but I think I am going to have to make do for 1.0 and open a follow on issue.

As I side note - the *start* of our focusing on accessibility is going to make it basically required that we have 1) core UI components that are well designed and accessible and 2) that core and third party extensions use them.

**Comment by jasongrout** [comment]
Even just fixing the dialog buttons (which IIRC are centrally located) would be a huge step forward, and would solve this issue as posted (i.e., the title is "for dialog buttons").

**Comment by ellisonbg** [comment]
I am coming to the conclusion that no one has figured out how to design good buttons that are usable and accessible. All the major web based design systems (Material Design, Blueprint, Fabric, etc.) all have a weird combinations of strengths and weaknesses across different states (focus, active, hover, etc.). Some observations:

* Background variations work well for hover and active states, because they are transient and are always accompanied by direct user interaction. As long as the light/dark/opacity of the background changes enough, it is accessible.
* However, background variations do not work well for focus, as a user doesn't always observe the transition to being focused. This is a problem Material Design suffers from as it uses a lighter background to indicate focus. If an UI appears with a button already focused, you have no idea that isn't just it's normal bg of the button. As a result, you have to tab around to observe changes in the bg colors to find focus. 
* The strongest indicator of focus is the blue border/outline. However, in a highly compact layout like JupyterLab, you can't really use an exterior offset outline, as is will often be cut off and clipped from adjacent elements and parent elements. I think the strongest approach for focus will be the 2px blue border (no inset of offset)–Figma uses this very effectively. The wonderful thing about this approach is that it naturally applies to just about any UI control you want to indicate focus on.
* However, the 2px solid border does not work well for bright solid filled buttons, which the dialogs have right now.
* As a compromise, the second best focus indicator is the inset white outline that Fabric uses. That works on solid buttons, so I am using it in the tentative fix. It is a compromise as it creates a more visually complex UI with a double border.
* However, the best long term solution will be to move away from solid buttons in general, and use the 2px blue border (in our brand color).

For reference, here is the Figma variation of the non-offset blue border for focus, that would well with flat, bordered, and text based buttons (not not solid color bgs):

![Screen Shot 2019-06-08 at 5 01 15 PM](https://user-images.githubusercontent.com/27600/59153469-10a67e00-8a0f-11e9-92d1-1646d022980a.png)

**Comment by blink1073** [comment]
Thanks for putting in all this research, @ellisonbg!

**Comment by ellisonbg** [comment]
Here is a design that meets the following goals and constraints:

* Covers normal, warn, and primary style buttons (we have these three types in the dialog).
* Covers active, hover, focus states.
* Meets WCAG minimum contrast ratio for text (4.5:1).
* No visual indicators outside the border of the button to prevent truncation in compact layouts.
* Stick to material design, 10 stop color palettes for each hue.

Light theme:

![Light Theme](https://user-images.githubusercontent.com/27600/59154775-3b083380-8a2f-11e9-859f-af5409a6cbef.jpg)

Dark theme:

![Dark Theme](https://user-images.githubusercontent.com/27600/59154838-59226380-8a30-11e9-8d5b-43096c7261c0.png)

The normal version here is almost identical to our toolbar buttons (with the addition of the bleu focus border). The downside of this approach is that the text color has to be darkened as the hover/active bg darkens to meet the contrast ratio.

Alternate versions possible:

* For the warn version, we could use the blue focus border, and white text on a red background.
* The hard one is the primary, where the blue border and a solid blue button would be indistinguishable.
  - We could pick an alternate primary color, but we have standardized on blue pretty strongly - lots would change.
  - The other good option from a brand perspective is the complementary orange, but that has a warning connotation that isn't compatible with a brand color.
  - Use a much lighter focus blue than the brand to create contrast.

**Comment by ellisonbg** [comment]
For 1.0 I have settled on the following (see #6517 for screenshots):

* For focus, use an outline of `1px solid var(--jp-brand-color1)` with offset of `4px` (exterior).
* Set `overflow:visible` on the containing `Widget` and only use this on dialog buttons where there is room (not on toolbar buttons).
* Slight opacity changes for hover and active states.
* Made dialog button labels caps case.

Longer term, we will need to dive into the design of our core UI components with an eye towards accessibility.
